### PR TITLE
Parallelize swift compilation by compiling file individually

### PR DESCRIPTION
**Pull request by nguyentruongtho**
This PR is one of the effort to rework the incremental build support https://github.com/facebook/buck/pull/907.

Instead of compiling all swift files in a target in one shot, we create build rule for each of the file so that we can compile them in parallel, therefore speed up the compilation process, especially useful for a target with a lot of swift files.

### CODE REVIEW

File: src/com/facebook/buck/swift/SwiftPrepareForCompile.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  private Step createFileListStep() {
+    ImmutableList<Arg> sourceListArgs = FileListableLinkerInputArg.from(
+        FluentIterable.from(SourcePathArg.from(getResolver(), srcs))
+            .filter(SourcePathArg.class)
+            .toList());
+
+    ProjectFilesystem fileSystem = getProjectFilesystem();
+    Path argFilePath = fileSystem.getRootPath().resolve(
```
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftPrepareForCompile.java]: we actually don't need argfile for swift, but I don't want to duplicate stuff by creating similar class like `CxxPrepareForLinkStep`.

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 2]
```diff
@@ -175,28 +228,32 @@ private SwiftCompileStep makeCompileStep() {
...
+          "-primary-file",
+          getResolver().getAbsolutePath(src.get()).toString(),
+          "-emit-module-path",
+          modulePath.toString(),
+          "-o",
+          objectPath.toString());
+    } else {
+      compilerCommand.add(
+          "-o",
+          modulePath.toString());
```
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftCompile.java]: When merging all single file outputs, the output is another swift module, not an object file.

**Comment by nguyentruongtho** [comment]
CI became crazy~

**Comment by nguyentruongtho** [comment]
interestingly, travis rebase your diff autoz and build with the rebased version, that's why it failed!

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 3]
```diff
@@ -265,15 +324,97 @@ public Path getPathToOutput() {
...
+              cxxPlatform,
+              swiftBuckConfig,
+              params.withFlavor(
+                  ImmutableFlavor.of(
+                      COMPILE_FLAVOR_PREFIX + outputName)),
+              sourcePathResolver,
+              swiftCompiler,
+              frameworkPaths,
+              moduleName,
+              outputPath,
```
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftCompile.java]: should we put all individual compiled file to the same folder like this, or use separate folders like cxx precompile?
I thought it could be a race condition when you compile `file1.swift` and `file2.swift` in parallel given `file2.swift` depending on `file1.swift`, but apparently compiling `file2.swift` doesn't result to compiling `file1.swift` (maybe it will do parsing & semantic analysis but won't actually compile `file1.swift`?). This is a really interesting topic and we need to understand more about swift compiler.

Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftCompile.java]: Use separate output directories. If you need them all in the same directory you will need to create a symlink view of them.

Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftCompile.java]: done. Also rebased to resolve conflict

File: src/com/facebook/buck/swift/SwiftPrepareForCompile.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import com.facebook.buck.step.Step;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.nio.file.Path;
+
+import javax.annotation.Nullable;
+
+class SwiftPrepareForCompile extends AbstractBuildRule {
```
Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftPrepareForCompile.java]: This doesn't need to be a separate build rule, it can be a step in the compilation rule, at least that's how we do it in all other cases.

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 5]
```diff
@@ -176,28 +230,32 @@ private SwiftCompileStep makeCompileStep() {
...
-    for (SourcePath sourcePath : srcs) {
-      compilerCommand.add(getResolver().getRelativePath(sourcePath).toString());
+
+    if (bridgingHeader.isPresent()) {
+      compilerCommand.add(
+          "-import-objc-header",
+          getResolver().getRelativePath(bridgingHeader.get()).toString());
+    }
+
+    if (src.isPresent()) {
```
Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftCompile.java]: It's really strange to have the rule behave so differently based on whether it has a source or not, can you split this into two rules?

File: src/com/facebook/buck/swift/SwiftPrepareForCompile.java (hunk 4 above)
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftPrepareForCompile.java]: no, coz the file list is used by all separate file build rules (dependencies of `SwiftPrepareForCompile` build target), better to create it upfront.

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 5 above)
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftCompile.java]: make sense, will do.

**Comment by nguyentruongtho** [comment]
Just rebased for huge java8 changes ^.

File: src/com/facebook/buck/swift/SwiftPrepareForCompile.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    super(buildRuleParams, resolver);
+    this.srcs = srcs;
+    this.linker = linker;
+    this.fileListPath = BuildTargets.getScratchPath(
+        getProjectFilesystem(), getBuildTarget(), "%s__filelist.txt");
+  }
+
+  private Step createFileListStep() {
+    ImmutableList<Arg> sourceListArgs = FileListableLinkerInputArg.from(
+        SourcePathArg.from(getResolver(), srcs).stream()
```
Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftPrepareForCompile.java]: Is this going to work for build target source paths?

If so, then absolute paths would be in the file list, so it shouldn't be cached (see https://github.com/facebook/buck/blob/master/src/com/facebook/buck/rules/SymlinkTree.java#L198-L205 for example).

If not, it should.

File: src/com/facebook/buck/swift/SwiftPrepareForCompile.java (hunk 4 above)
Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftPrepareForCompile.java]: I see, makes sense.

**Comment by Coneko** [comment]
Back to you for the rule split.

File: src/com/facebook/buck/cxx/CxxCreateArgFileStep.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2014-present Facebook, Inc.
```
Comment (Coneko) [inline src/com/facebook/buck/cxx/CxxCreateArgFileStep.java]: 2016

File: src/com/facebook/buck/cxx/CxxCreateFileListStep.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2014-present Facebook, Inc.
```
Comment (Coneko) [inline src/com/facebook/buck/cxx/CxxCreateFileListStep.java]: 2016

File: src/com/facebook/buck/cxx/CxxCreateArgFileStep.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+  private final Path argFilePath;
+  private final Path output;
+  private final ImmutableList<Arg> args;
+  private final Linker linker;
+
+  public CxxCreateArgFileStep(
+      Path argFilePath,
+      Path output,
+      ImmutableList<Arg> args,
```
Comment (Coneko) [inline src/com/facebook/buck/cxx/CxxCreateArgFileStep.java]: It's really strange that this filters out the file list args.

File: src/com/facebook/buck/cxx/CxxCreateArgFileStep.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    this.args = args;
+    this.linker = linker;
+  }
+
+  private void createArgFile() throws IOException {
+    if (Files.notExists(argFilePath.getParent())) {
+      Files.createDirectories(argFilePath.getParent());
+    }
+    LOG.verbose("Cleaning up previous argfile at " + argFilePath.toString());
+    Files.deleteIfExists(argFilePath);
```
Comment (Coneko) [inline src/com/facebook/buck/cxx/CxxCreateArgFileStep.java]: No other step deletes its output before writing it, remove this.

File: src/com/facebook/buck/cxx/CxxCreateFileListStep.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    this.fileListPath = fileListPath;
+    this.args = args;
+  }
+
+  private void createFileList() throws IOException {
+    if (Files.notExists(fileListPath.getParent())) {
+      Files.createDirectories(fileListPath.getParent());
+    }
+    LOG.verbose("Cleaning up previous filelist at " + fileListPath.toString());
+    Files.deleteIfExists(fileListPath);
```
Comment (Coneko) [inline src/com/facebook/buck/cxx/CxxCreateFileListStep.java]: Same as above.

**Comment by Coneko** [comment]
Rather than having `CxxCreateArgFileStep` and `CxxCreateFileListStep` encode so much information about the specifics of their use case, making them very unintuitive to use, merge them into a single class that is more reusable and behaves more like a normal step: don't delete the output before writing out the new one, and pass in an optional escaper to use.

Also you still have to address the issue of splitting `SwiftCompile`.

**Comment by nguyentruongtho** [comment]
> Rather than having CxxCreateArgFileStep and CxxCreateFileListStep encode so much information about the specifics of their use case, making them very unintuitive to use, merge them into a single class that is more reusable and behaves more like a normal step: don't delete the output before writing out the new one, and pass in an optional escaper to use.

> Also you still have to address the issue of splitting SwiftCompile.

I preserved `SwiftPrepareForCompile` as a composite step of the other two, one has to follow the other in the right order. Not quite follow the deletion suggestion, I thought we did the same thing in `CxxPrepareForLinkStep`? 

I think creating filelist step is worth to be taken out since it is going to be used in both linking (`CxxLink`) and compiling (`SwiftCompile`), maybe we can remove `CxxCreateArgFileStep`, but it's not clearly better though.

Any suggestion for splitting the step?

**Comment by Coneko** [comment]
I mean add a step that takes a list of args, a path to a file and an optional escaper, and writes out the args to the file optionally using the escaper, and change the code to use this generic step instead of having the two new steps.

This should be done in a separate PR, then rebase this PR on that one.

**Comment by nguyentruongtho** [comment]
hmm, but for `SwiftCompile`, I don't need the args file, `CxxCreateFileListStep` is what I actually need. We don't want to duplicate the logic for creating file list in both `CxxCreateFileListStep` and `CxxPrepareForLinkStep`, that's why I want to split it into 2 steps.

So, if we want to keep `CxxPrepareForLinkStep` class, we need to rename it, since it is going to be used for `SwiftCompile` also, agree?

**Comment by Coneko** [comment]
It's fine to split it into two steps, I'm saying make it two instances of the same step class, instead of making two different step classes.

**Comment by nguyentruongtho** [comment]
Brilliant! Didn't notice that we could make one common step out of the two.

**Comment by Coneko** [comment]
This looks good but is blocked on the test failure in #944.

**Comment by Coneko** [comment]
Back to you to rebase after #944 lands.

**Comment by nguyentruongtho** [comment]
@Coneko Rebased. It should pass the test now but plz don't import it yet. I've just found an edge case when this fails. I'll fix it and add more test soon.

**Comment by nguyentruongtho** [comment]
Pushed a wrong branch, just force pushed again with the right one. But the failing edge case is still not handled.

**Comment by nguyentruongtho** [comment]
ok, let's come back to this mess 😿

**Comment by Coneko** [comment]
Seems like it's not merging cleanly.

**Comment by steeve** [comment]
Btw, I'm actually using a custom buck build from this PR, and it works great !
Hopefully this will get merged soon

**Comment by nguyentruongtho** [comment]
Not going to work more further on this, as we had some benchmarking @Uber and come to a conclusion that building one target in one shot is much faster than building file by file.

**Comment by steeve** [comment]
We had the exact opposite experience. Compiling file by file is much faster (and also incremental).
We can open a new PR if you guys want (we made some modifications).

**Comment by ultramiraculous** [comment]
With a large mixed ObjC/Swift project we had a huge (Around 60%) speedup from building the target all at once and turning on whole module optimization.

**Comment by nguyentruongtho** [comment]
that's very interesting result, can you put up a branch with your modifications?

**Comment by daedric** [comment]
here you go :)
https://github.com/daedric/buck/tree/swift-zenly-rebased

**Comment by steeve** [comment]
We also have some CocoaPods helpers (made in python) to parse and use pods from inside a BUCK file.

**Comment by nguyentruongtho** [comment]
Nice, thanks @daedric and @steeve I'll check if we can get a significant improvement on our side, if yes, we should reopen this and work to get it landed together :).

**Comment by daedric** [comment]
That would be great, do not hesitate asking me question. My branch is unfortunately not clean as it was a bit rushed, but I'm willing to provide any assistance you need to clean it

cheers,

**Comment by gbrhaz** [comment]
@steeve I'd be interested in seeing those CocoaPod scripts - looking at doing this myself. Any chance of putting them up?

**Comment by nguyentruongtho** [comment]
I had a look through https://github.com/daedric/buck/tree/swift-zenly-rebased, there is not a lot of changes comparing to this original pr, I'm glad that it works for your project. 

I think this approach would work for a large target (many files) with a lot of separate parts in it, since in this case, there shouldn't be lot of dependencies between the files/classes, and buck will be able to parallelize the jobs better when compiling per-file targets. I would ague that if this situation happens, we should refactor the big target and split it into smaller targets. However, this approach won't benefit much in general, and doesn't guarantee that it will be always faster, at least in our case, it is slower than compiling per-target with [whole module optimization](https://swift.org/blog/whole-module-optimizations/) enabled.

We found a better way of doing this and will not continue to support this pr anymore, but we will revisit supporting better parallelization compilation for swift later.
